Tanner p1

IsaacDerg opened the pull request:
Completed my parts of the program.

IsaacDerg committed:
added functionality to set destructor

IsaacDerg committed:
added remove_current functionality

added and commented remove_current function code

IsaacDerg committed:
added notes

added notes to Set deconstructor

IsaacDerg committed:
added insert functionality

added insert functionality and comments

IsaacDerg committed:
edited comment

IsaacDerg committed:
added check in insert function

added check for existing data

IsaacDerg committed:
added functionality to search

added functionality and notes to search function

IsaacDerg committed:
updated search function

used temp to search and only changed current pointer if found

IsaacDerg committed:
Merge branch 'main' into tanner-p1

IsaacDerg committed:
updated insert

current now points to the new node if created

IsaacDerg committed:
added constructor functionality

IsaacDerg committed:
combined two statements and added a note

lavender-aa commented on program-1/set.cpp:
```diff
@@ -3,31 +3,106 @@
+
+        // delete current node
+        delete current;
+    }
```
this leaves `current` dangling (just assign to nullptr after loop).

not sure if this would actually ever be an issue since `current` would be immediately going out of scope, but it's good practice

lavender-aa commented on program-1/set.cpp:
```diff
@@ -3,31 +3,106 @@
+    current = head;
+
+    // loop through the list until the next node data is greater than num and the next node is not null
+    while(current->data < num && current->next != nullptr)
```
segfault:
on first insert, `head` is null, results in null dereference (`current->data`)

also, the comment is misleading (loops until num is greater or equal to current node num or current is the last node)

lavender-aa commented on program-1/set.cpp:
```diff
@@ -3,31 +3,106 @@
+        Node* temp = new Node;
+
+        // set new node data
+        temp->data = num;
```
can create and initialize in one line, since the `Node` struct has a constructor 
(`*Node temp = new Node(num);`)

IsaacDerg replied:
I tried this, but the intelliSense didn't like when I did this. I get `no instance of constructor "Node::Node" matches the argument list`.

lavender-aa replied:
i've figured out why; when i was making this project i forgot to include the parameter on the node constructor (currently it's signature is just `Node()` instead of `Node(int num)`). might be good to have both a default and a parameterized one imo but it's up to you

IsaacDerg replied:
Ohh that makes sense. That's what I figured was the issue. I'll go ahead and add a parameterized one.

lavender-aa commented on program-1/set.cpp:
```diff
@@ -3,31 +3,106 @@
 int Set::remove_current()
 {
+    // variables
+    Node* temp = current->next;
```
possible segfault (null deref) if there are no nodes (for example, if the list was just initialized)

lavender-aa commented on program-1/set.cpp:
```diff
@@ -3,31 +3,106 @@
+    int val = current->data;
+
+    // get data from next node
+    current->data = current->next->data;
```
possible segfault (null deref) if `current` is tail node.

also, while not an error, you may as well use `temp->data` since you already set it to `current->next`

lavender-aa commented on program-1/set.cpp:
```diff
@@ -3,31 +3,106 @@
+    current->next = temp->next;
 
+    // delete the node
+    delete temp;
```
`temp` left dangling

lavender-aa reviewed (changes requested):
spotted some things that need to be addressed. might not be everything, but a decent amount.

if you don't have a debugger set up, i can help you get it going; it makes testing much easier (which i assume you didn't do, since calling `insert` always segfaults the first time)

IsaacDerg commented:
I would greatly take the help setting up a debugger. Can I just debug set.cpp file or do I need to call the functions in main when debugging? Sorry if that seems like a dumb question.

lavender-aa commented:
first, you need to make sure you have gdb installed (which you probably do, but it doesn't hurt to check). just run `gdb --version` in a console

with vscode, you typically set up a debugger for each individual project. i'm sure this can be automated, but here's how it works:
- in your project source directory, you create a hidden folder called `.vscode`
- inside this folder, you need two files: `tasks.json` and `launch.json`

whenever you go to debug, you'll be able to choose the task (typically called build) from your `launch.json`.

`launch.json` contains lots of important information required to debug a project (it would be a lot easier if this was just a single file), one of which is the task it should run before trying to debug a program. this task is typically one that builds your project into a binary with the debugger flags set on your compiler

easiest way for me to get you the config for this repo would probably just be by email, so i'll do that

lavender-aa commented:
oh actually even easier: those two files were commited to the old [object oriented repo](https://github.com/lavender-aa/oop-groupwork/tree/main/.vscode), so you can copy them from there. it's possible you'll need to change the `main.bin` to `main.exe` in launch, though

lavender-aa commented:
oh i misunderstood your question

yeah, you have to test your functions by putting breakpoints inside of them and calling them in main

IsaacDerg commented:
Thanks, let me test my changes.

IsaacDerg commented:
I set the `launch.json` and `tasks.json` up and it's not working. I remember you got it to work for me last semester, but I can't remember what it was. It keeps going to the terminal and erroring out on executing task.

lavender-aa commented:
probably way too hard to try and figure out what's wrong here, if you want/are able to i work 11-1 tomorrow in noss 110 and could help you there

IsaacDerg commented:
I can come in, I'll try to be there around 11. For now I'll just commit the changes I've made.

IsaacDerg committed:
applied required changes

fixed a few dangling pointers, made remove_current account for an empty list or current is the tail, made insert account for an empty list, fixed some comments

lavender-aa commented:
whenever you do change your code according to a review you can click the resolve button

IsaacDerg commented:
Ah, I forgot about that, thanks.

IsaacDerg committed:
added parameterized constructor for node

added parameterized constructor for node and combined two statements

lavender-aa commented on program-1/set.cpp:
```diff
@@ -3,31 +3,154 @@
+    Node* temp = head;
+
+    // loop through the list if we did not reach the end of the list and temp data is not equal to num
+    while(temp->next != nullptr && temp->data != num)
```
segfault if there are no nodes

lavender-aa commented on program-1/set.cpp:
```diff
@@ -3,31 +3,154 @@
+            Node* prev = head;
+
+            // loop through to get the node before current
+            while(prev->next != current)
```
segfault if current = head; condition inside while will always be true in this case, leading to prev = nullptr and subsequent null deref

lavender-aa reviewed (changes requested):
few more things found with the help of the debugger

IsaacDerg committed:
added check to search

search now checks if the list is empty

IsaacDerg committed:
updated remove_current

hopefully should work now, debugged by drawing it out and seems to work for all scenarios of a list

lavender-aa reviewed (approved):
did some testing and it seems to work as expected. the only real problem is that <iostream> isn't included in `set.h`, but that doesn't have to be addressed here (the line is included in my branch, so once that's merged in it'll be there)

lavender-aa commented:
i'll merge now, if problems come up later you can make changes on your branch and reopen the PR

lavender-aa commented:
after merging and testing, something have been brought to my attention:

the first is that `insert()` doesn't work properly. some examples; left side is number added, right side is the list:
```
1:     1 -> null
0:     1 ->  0  -> null
-1:    1 -> -1 -> 0 -> null
```
```
-1:   -1 -> null
0:    -1 ->  0  -> null
-2:   -1 -> -2 -> 0 -> null
```

just two examples, but you get the point (and you can test on your own)

i'm assuming this is because `insert()` only ever inserts a new node after the while loop ends. might need to refactor the logic

IsaacDerg commented:
Alright I'll take a look at it now.
